python/hsfs/core/tfdata_engine.py

```diff
@@ -0,0 +1,484 @@
+        """
+
+        if self._training_dataset_format.lower() not in ["tfrecords", "tfrecord"]:
+            raise Exception(
+                "tf_record_dataset function works only for training datasets that have tfrecord or tfrecords format"
+            )
+
+        if process and (batch_size is None or num_epochs is None):
+            raise ValueError(
+                "if process is set to True you also need to provide batch_size and num_epochs"
+            )
+
+        if one_hot_encode_labels and (num_classes is None or num_classes <= 1):
+            raise ValueError(
+                "if one_hot_encode_labels is set to True you also need to provide num_classes > 1"
+            )
+
+        dataset, tfrecord_feature_description = _get_tfdataset(
+            self._input_files, self._cycle_length
+        )
+
+        def _de_serialize(serialized_example):
+            example = tf.io.parse_single_example(
+                serialized_example, tfrecord_feature_description
+            )
+            return example
+
+        def _process_example(example):
+            # get target variable 1st
+            y = example[self._target_name]
+            if one_hot_encode_labels:
+                y = tf.one_hot(y, num_classes)
+            else:
+                y = tf.cast(y, tf.float32)
+
+            # if there is only 1 feature we return it
+            if len(self._feature_names) == 1:
+                _feature_name = self._feature_names[0]
+                # here it is assumed that if user provides serialized_ndarray, it is only one feature
+                if self._feature_names in serialized_ndarray_fname:
+                    x = tf.io.parse_tensor(example[_feature_name], out_type=tf.float32)
+                else:
+                    x = example[_feature_name]
+                return x, y
+            # Otherwise we need to have features in the same type, thus tf.float32
+            else:
+                x = []
+                for _feature_name in self._feature_names:
+                    x.append(_convert2float32(example[_feature_name]))
+                x = tf.stack(x)
+            return x, y
+
+        dataset = dataset.map(
+            lambda value: _de_serialize(value),
+            num_parallel_calls=tf.data.experimental.AUTOTUNE,
+        )
+
+        if process:
+            dataset = dataset.map(
+                lambda value: _process_example(value),
+                num_parallel_calls=tf.data.experimental.AUTOTUNE,
+            )
+            dataset = _optimize_dataset(
+                dataset, batch_size, num_epochs, self._is_training
+            )
+
+        return dataset
+
+    def tf_csv_dataset(
+        self,
+        batch_size=None,
+        num_epochs=None,
+        one_hot_encode_labels=False,
+        num_classes=None,
+        process=False,
+    ):
+        """
+        Reads csv files and returns CsvDatasetV2 or PrefetchDataset object ready to feed tf keras models
+
+        Example:
```

Comment:
see my other comment on how to write examples. if you write it like this, it will look like this

![image](https://user-images.githubusercontent.com/8422705/95957886-9a830a00-0e00-11eb-80b6-974ef5b0c17b.png)


![image](https://user-images.githubusercontent.com/8422705/95957787-758e9700-0e00-11eb-8fe7-49cc2864fe7d.png)